src/libopenrave/libopenrave.h

```diff
@@ -481,6 +481,145 @@ void UpdateOrCreateInfo(const rapidjson::Value& value, std::vector<boost::shared
+                if (nameIt != infoIdxByName.end() && nameIt->second == existingInfoIndex) {
+                    infoIdxByName.erase(nameIt);
+                }
+
+                // Don't erase the backing item from the list immediately, since that would invalidate all the subsequent indices in our jump tables.
+                // Instead just null the element, and we'll filter the whole list at the end.
+                vInfosOut[existingInfoIndex].reset();
+                didEraseElements = true;
+                continue;
+            }
+
+            // If we are updating an element in place, we need to check if we're about to change the id/name
+            // If we are, we have to update the jump tables.
+            const bool didChangeName = !name.empty() && vInfosOut[existingInfoIndex]->_name != name; // If the name to be loaded is empty, it won't clobber anything, so don't count that as a change
+            const bool didChangeId = vInfosOut[existingInfoIndex]->_id != id; // Id always overwrites, so count it as changed even if the new id is empty
+
+            // Erase the old id/name mapping if it changed
+            if (didChangeId) {
+                infoIdxById.erase(vInfosOut[existingInfoIndex]->_id);
+            }
+            if (didChangeName) {
+                infoIdxByName.erase(vInfosOut[existingInfoIndex]->_name);
+            }
+
+            // Deserialize the new json over the existing entry
+            vInfosOut[existingInfoIndex]->DeserializeJSON(value, fUnitScale, options);
+            vInfosOut[existingInfoIndex]->_id = id;
+
+            // Write back the new id/name mappings with the new values if changed
+            if (didChangeId) {
+                infoIdxById.emplace(vInfosOut[existingInfoIndex]->_id, existingInfoIndex);
+            }
+            if (didChangeName) {
+                infoIdxByName.emplace(vInfosOut[existingInfoIndex]->_name, existingInfoIndex);
+            }
+
+            continue;
+        }
+
+        // If the item is deleted and didn't match an existing index, then there is nothing to remove.
+        // Just skip this entry.
+        if (isDeleted) {
+            continue;
+        }
+
+        // If we didn't match an existing entry, then create a new one now
+        boost::shared_ptr<T> pNewInfo(new T());
+        pNewInfo->DeserializeJSON(value, fUnitScale, options);
+        pNewInfo->_id = id;
+
+        // Update our jump tables and push this info onto the list
+        const size_t newInfoIndex = vInfosOut.size();
+        infoIdxById.emplace(id, newInfoIndex);
+        infoIdxByName.emplace(pNewInfo->_name, newInfoIndex);
+        vInfosOut.emplace_back(std::move(pNewInfo));
+    }
+
+    // If we deleted any elements, scan through and ensure there are no holes in the output before returning.
+    // Scan backwards through the list, swapping all of the invalid pointers to the end.
+    // Once we're done, truncate the list to drop all the invalid entries.
```

Comment:
I delete things this way to avoid `memmove`, but the tradeoff is that we don't guarantee the total ordering of links with this implementation - is this something that is implicitly relied on somewhere? I think the old impl would have the same issue if you had a delete + add, but wanted to bring it up to be sure.

Reply:
> I think the old impl would have the same issue if you had a delete + add

I think old information (`UpdateOrCreateInfoWithNameCheck`) doesn't change the order of elements in `vInfos`, does it?

> but the tradeoff is that we don't guarantee the total ordering of links with this implementation - is this something that is implicitly relied on somewhere? 

I cannot think of this assumption being made somewhere, but I feel that it is safer to keep the order. What do you think @rdiankov ?

Reply:
Changing order may not be welcome when one has to look at diff of KinBodyInfo JSON.
This will be order preserving and still linear time. Should be better than the old method. can we do this?
```c++
size_t validElementCount = 0;
for (ssize_t iInfo = 0; iInfo < vInfosOut.size(); ++iInfo) {
    if( !!vInfosOut[iInfo] ) {
        if( validElementCount != iInfo ) {
            vInfosOut[validElementCount] = vInfosOut[iInfo];
        }
        ++validElementCount;
    }
}
vInfosOut.resize(validElementCount);
```